Downstream build support

TestYourLuck opened the pull request:
Commit status updater usually takes project id from build cause if it is caused by GitLabWebHookCause, but when you want to update status from downstream job, you can't due that. This improvement changes that.
The other way you could be passing all needed parameters to downstream job with a bit of extra configuration, but it also does not work, due to #538 bug.
Added test for new code. Also all old tests are passing.
Additionally added log entry when project id fails to be resolved.

Commit:
Added support for GitLab status updating from downstream builds where upstream is caused by webhook

Commit:
Updated test for GitLab status updating from downstream builds where upstream is caused by webhook

Commit:
Updated CommitStatusUpdater to support any depth upstream webhook trigger cause

Commit:
Updated test for GitLab status updating from downstream builds where few level upstream is caused by webhook

TestYourLuck commented:
As for now, ugly workaround is to update commit status manually from groovy (assuming job is triggered by GitLabWebHook):
```
def updateCommitStatus(String state, BuildState status) {
    GitLabApi client = getClient(manager.build)
    try {
        client.getCommit(env.gitlabProjectId, env.gitlabCommitHash)
        echo 'Commit to update found.'
        client.changeBuildStatus(env.gitlabProjectId, env.gitlabCommitHash, status, env.gitlabBranch, state, env.fullBuildUrl, null);
    } catch (NotFoundException e) {
        echo 'Commit to update not found!'
    } catch (WebApplicationException | ProcessingException | IOException | InterruptedException | IllegalStateException e) {
        echo 'Failed to update GitLab commit status'
    }
}

def init() {
    env.fullBuildUrl = Jenkins.getInstance().getRootUrl() + manager.build.getUrl()

    for (cause in manager.build.causes) {
        if (cause instanceof UpstreamCause) {
            for (upcauses in cause.getUpstreamCauses()) {
                if (upcauses instanceof GitLabWebHookCause) {
                    env.gitlabProjectId = upcauses.getData().getSourceProjectId().toString()
                    env.gitlabBranch = upcause.getData().getSourceBranch()
                    break
                }
            }
            break
        }
    }
    BuildData buildData = manager.build.getAction(BuildData.class)
    Revision lastBuiltRevision = buildData.getLastBuiltRevision()
    env.gitlabCommitHash = buildData.getLastBuild(lastBuiltRevision.getSha1()).getMarked().getSha1String()
}

node() {
  updateCommitStatus('jenkins', BuildState.success)
}
```

omehegan commented:
@TestYourLuck can you resolve the merge conflicts?

Commit:
Merge remote-tracking branch 'default/master' into downstream_build_support

# Conflicts:
#	src/main/java/com/dabsquared/gitlabjenkins/util/CommitStatusUpdater.java

TestYourLuck commented:
@omehegan Done

Commit:
Fixed test

omehegan commented:
@mreichel @colourmeamused can you review?

omehegan commented:
@TestYourLuck blah, merge conflicts again, sorry.

mreichel reviewed (approved):
I think the pr is fine, only the merge conflicts need to be resolved

Commit:
Merge remote-tracking branch 'default/master' into downstream_build_support

# Conflicts:
#	src/main/java/com/dabsquared/gitlabjenkins/util/CommitStatusUpdater.java

TestYourLuck commented:
Fixed merge conflicts.
Added few java comments about strange code from master - please look into it.
Fixed poor spacing that came from master branch - tabs and spaces mixed in same line all over changed code (worth to mention file originally was spaces only).
@mreichel please review

omehegan commented:
@TestYourLuck sorry, I think I just introduced more merge conflicts here. I am working on doing a plugin release today or tomorrow, if you can resolve I'll get this merged. Thanks!

TestYourLuck commented:
I'm on vacation, but I'll see what I can do.

TestYourLuck committed:
Merge branch 'master' into downstream_build_support

mreichel commented:
The changes look fine to me! Thank you.

omehegan commented:
@TestYourLuck if I am understanding right, #595 is also supposed to accomplish this result. Can you take a look at that PR and confirm?

TestYourLuck commented:
Looks similar, but I think that MergeRequestNotifier is used elsewhere. I can't confirm that since I can't search the code on my mobile. And I will not be able to do that for whole weekend. @omehegan, could you or someone else check if that andmy changed classes used in same place? If not, then its probably different changes.

omehegan commented:
@TestYourLuck now that we have merged API v4 support I am getting back to the other open PRs. can you fix these failing tests?

mreichel committed:
Merge remote-tracking branch 'upstream/master' into downstream_build_support

mreichel committed:
fix tests

mreichel committed:
Merge remote-tracking branch 'upstream/master' into downstream_build_support

mreichel committed:
fix tests after commit f56a896039052f55dbec24518d9f4c248ec21b56

mreichel commented:
@omehegan tests are fixed

Argelbargel commented on src/test/java/com/dabsquared/gitlabjenkins/util/CommitStatusUpdaterTest.java:
```diff
@@ -116,7 +117,7 @@ public void setUp() throws Exception {
 		CommitStatusUpdater.updateCommitStatus(build, taskListener, BuildState.success, STAGE);
 		
-		verify(client).changeBuildStatus(Integer.toString(PROJECT_ID), REVISION, BuildState.success, null, STAGE, Jenkins.getInstance().getRootUrl() + BUILD_URL, null);
+		verify(client).changeBuildStatus(Integer.toString(PROJECT_ID), REVISION, BuildState.success, null, STAGE, Jenkins.getInstance().getRootUrl() + BUILD_URL + BUILD_URL_REDIRECT, null);
```
I think using DisplayURLProvider.get().getRunURL(build) as in GitLabCommitStatusPublisherTest would be even better...

mreichel replied:
That is correct! I fixed it.

mreichel committed:
fix review issue

TestYourLuck commented:
Finally, thanks guys for merging :)

omehegan commented:
This has just been released in plugin version 1.5.2. It will be in the update center within a few hours.
